### PR TITLE
0.3.10

### PR DESCRIPTION
**Pull request by claudiolemos**
- Removed unecessary dependencies
- All providers except openai now work with REST API requests, making the packcage much lighter
- Tool calling with Llama 3.1 using Azure
- Tool calling with Gemini models using Vertex

### CODE REVIEW

File: llmstudio/engine/providers/anthropic.py [hunk 1]
```diff
@@ -39,35 +40,62 @@ async def generate_client(
...
+            return await asyncio.to_thread(
+                requests.post,
+                url,
+                headers=headers,
+                json=data,
+                stream=True,
             )
-        except anthropic._exceptions.APIError as e:
-            raise HTTPException(status_code=e.status_code, detail=e.response.json())
+        except Exception as e:
```
Comment (diogoncalves) [inline llmstudio/engine/providers/anthropic.py:69]: what if the request's result is an error? ie. status_code != 200?

File: llmstudio/engine/providers/anthropic.py [hunk 2]
```diff
@@ -76,16 +104,6 @@ async def parse_response(
...
-                        id=str(uuid.uuid4()),
-                        choices=[
-                            Choice(delta=ChoiceDelta(), finish_reason="stop", index=0)
-                        ],
-                        created=int(time.time()),
-                        model=kwargs.get("request").model,
-                        object="chat.completion.chunk",
-                    ).model_dump(),
+                    ).model_dump()
+                except:
```
Comment (diogoncalves) [inline llmstudio/engine/providers/anthropic.py:108]: this is dangerous, which errors should we explicitely allow to pass, and which ones should we raise?

File: llmstudio/engine/providers/azure.py [hunk 3]
```diff
@@ -23,7 +33,7 @@ class AzureRequest(ChatRequest):
     api_version: Optional[str] = None
     base_url: Optional[str] = None
     parameters: Optional[AzureParameters] = AzureParameters()
-    functions: Optional[List[Dict[str, Any]]] = None
+    tools: Optional[List[Dict[str, Any]]] = None
```
Comment (diogoncalves) [inline llmstudio/engine/providers/azure.py:36]: this requires changes on the clients using function calling with azure openai models currently, right?

File: llmstudio/engine/providers/azure.py [hunk 4]
```diff
@@ -44,48 +56,284 @@ async def generate_client(
...
                 )
-                return await asyncio.to_thread(
-                    client.chat.completions.create,
-                    model=request.model,
-                    messages=(
-                        [{"role": "user", "content": request.chat_input}]
-                        if isinstance(request.chat_input, str)
-                        else request.chat_input
-                    ),
-                    functions=request.functions,
```
Comment (diogoncalves) [inline llmstudio/engine/providers/azure.py:77]: should we keep both "functions" and "tools" logging an deprecation warning for those using functions?

File: llmstudio/engine/providers/azure.py [hunk 5]
```diff
@@ -44,48 +56,284 @@ async def generate_client(
...
+        self, response: AsyncGenerator, **kwargs
+    ) -> AsyncGenerator[str, None]:
+        saving = False
+        function_name = ""
+        function_response = False
         for chunk in response:
-            yield chunk.model_dump()
+            content = chunk.choices[0].delta.content
+            if content is not None:
+                if "§" in content:
```
Comment (diogoncalves) [inline llmstudio/engine/providers/azure.py:131]: shouldn't it be a regex? what if the function has the "section sign" in it's call?

`# Example dataset
data = {
    'feature1': np.random.rand(100),
    'feature2': np.random.rand(100),
    'feature3': np.random.rand(100),
    '§': np.random.rand(100)  # Target variable with section sign
}
`

File: llmstudio/engine/providers/azure.py [hunk 6]
```diff
@@ -44,48 +56,284 @@ async def generate_client(
...
+        if isinstance(message, str):
+            return [{"role": "user", "content": message}]
+        return message
+
+    def add_system_message(self, openai_message, llama_message, tools):
+        system_message = ""
+        system_message_found = False
+        for message in openai_message:
+            if message["role"] == "system" and message["content"] is not None:
+                system_message_found = True
```
Comment (diogoncalves) [inline llmstudio/engine/providers/azure.py:255]: this flag will never be False again. is it compromising the condition on line 260?

File: llmstudio/engine/providers/openai.py [hunk 7]
```diff
@@ -48,6 +50,7 @@ async def generate_client(
                     if isinstance(request.chat_input, str)
                     else request.chat_input
                 ),
+                # tools=request.tools,
```
Comment (diogoncalves) [inline llmstudio/engine/providers/openai.py:53]: so, if the client passes tools, it is ignored? shouldn't we use tools if tools are provided and functions if functions are provided and tools is None?

File: llmstudio/engine/providers/provider.py [hunk 8]
```diff
@@ -182,52 +185,58 @@ async def handle_response(
...
-                                        type=tool_call_type,
-                                    )
-                                ],
-                            ),
-                        )
-                    ],
-                ),
-                tool_call_arguments,
+            tool_call_type = tool_calls[0].get("function").get("type")
+            # tool_call_type = "function"
```
Comment (diogoncalves) [inline llmstudio/engine/providers/provider.py:200]: remove

File: llmstudio/engine/providers/provider.py [hunk 9]
```diff
@@ -253,7 +264,9 @@ def join_chunks(self, chunks, request):
                     part = chunk.get("arguments", "")
                     if part:
                         function_call_arguments += part
-                elif isinstance(request, OpenAIRequest):
+                elif isinstance(request, OpenAIRequest) or isinstance(
```
Comment (diogoncalves) [inline llmstudio/engine/providers/provider.py:267]: do we need this condition? we're already with a request of instances OpenAIRequest or VertexAIRequest as defined on line 251

File: llmstudio/engine/providers/provider.py [hunk 10]
```diff
@@ -253,7 +264,9 @@ def join_chunks(self, chunks, request):
                     part = chunk.get("arguments", "")
```
Comment (diogoncalves) [inline llmstudio/engine/providers/provider.py:264]: is this code ever running? how could we have an AzureRequest if we're in the scope of the condition of line 251? ie. `            elif isinstance(request, OpenAIRequest) or isinstance(
                request, VertexAIRequest
            ):`

File: llmstudio/engine/providers/provider.py [hunk 11]
```diff
@@ -392,10 +405,8 @@ def output_to_string(self, output):
     def get_end_token_string(self, metrics: Dict[str, Any]) -> str:
         return f"{self.END_TOKEN},input_tokens={metrics['input_tokens']},output_tokens={metrics['output_tokens']},cost_usd={metrics['cost_usd']},latency_s={metrics['latency_s']:.5f},time_to_first_token_s={metrics['time_to_first_token_s']:.5f},inter_token_latency_s={metrics['inter_token_latency_s']:.5f},tokens_per_second={metrics['tokens_per_second']:.2f}"
 
-    def _get_tokenizer(self) -> Tokenizer:
-        return {
-            "anthropic": Anthropic().get_tokenizer(),
-        }.get(self.config.id, tiktoken.get_encoding("cl100k_base"))
+    def _get_tokenizer(self):
+        return {}.get(self.config.id, tiktoken.get_encoding("cl100k_base"))
```
Comment (diogoncalves) [inline llmstudio/engine/providers/provider.py:409]: `self.config.id` key will never exist in the empty dictionary `{}`

File: llmstudio/engine/providers/vertexai.py [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+                    id=str(uuid.uuid4()),
+                    choices=[
+                        Choice(delta=ChoiceDelta(), finish_reason="stop", index=0)
+                    ],
+                    created=int(time.time()),
+                    model=kwargs.get("request").model,
+                    object="chat.completion.chunk",
+                ).model_dump()
+
+    def convert_openai_to_vertexai(self, input_data, tool_payload):
```
Comment (diogoncalves) [inline llmstudio/engine/providers/vertexai.py:260]: let's discuss this method offline

File: llmstudio/llm/__init__.py [hunk 13]
```diff
@@ -249,14 +268,23 @@ async def async_stream(self, input: str, **kwargs):
...
+            try:
+                error_data = response.json().get("detail", "LLMstudio Engine error")
+            except ValueError:
+                error_data = response.text
+            raise Exception(error_data)
+
+        for chunk in response.iter_content(chunk_size=None):
+            if chunk:
+                yield chunk.decode("utf-8")
+                # yield ChatCompletionChunk(**chunk.decode("utf-8"))
```
Comment (diogoncalves) [inline llmstudio/llm/__init__.py:290]: do we need this comment?

File: llmstudio/tests/__init__.py
Comment (diogoncalves) [inline llmstudio/tests/__init__.py:1]: why deleting all the tests?

**Review by diogoncalves — CHANGES_REQUESTED** [review]

File: llmstudio/engine/providers/azure.py (hunk 3 above)
Comment (diogoazevedo15) [reply llmstudio/engine/providers/azure.py:36]: You're absolutely right. I'm currently working on a fix to revert that change.

File: llmstudio/engine/providers/azure.py (hunk 4 above)
Comment (diogoazevedo15) [reply llmstudio/engine/providers/azure.py:77]: I think this would be a good approach to the problem.

**Pull request merged**